### DISCUSSION TITLE
Possible reasons for incorrect output

### DISCUSSION THREAD

**Discussion by jinghuiF** [brandon-rhodes/python-sgp4#101] [Q&A]
Hello, I'm trying to use the library to propagate GPS satellite's position, but the output I get is not corresponding with the real position. Could you help me out if I made any mistakes?

I want to calculation PRN13 satellite position in 2021-11-8 22:45:00 UTC.

```
from sgp4.api import jday, Satrec

 # GPS BIIR-2  (PRN 13)
longstr1 = '1 24876U 97035A   21340.91043198  .00000063  00000+0  00000+0 0  9993'
longstr2 = '2 24876  55.4877 163.1850 0055575  54.7069 305.8344  2.00562886178793'

jd, fr = jday(2021,11,8,22,45,0)
satellite = Satrec.twoline2rv(longstr1, longstr2)
_, txPos, txV = satellite.sgp4(jd, fr)

print('Position: {}'.format(txPos))
```
The output I get from the code is `Position: (-19961.109688580247, 13546.39913549107, -11089.456691307081)`, which is really different from the real position `(-17632.902569, 19792.732070, -1400.735849)`.  Both are coordinates in TEME.

**Sorry I just found the weird part of the plot is because of one mistake within the code. 
But the disagreement of the answers shown above still remains.** 

![Figure 2022-03-03 103411](https://user-images.githubusercontent.com/87004784/156479241-5454ab19-03d1-42bd-97fc-31cfb46cdce0.png)

```
import numpy as np
from sgp4.api import jday, Satrec

# (2021, 11, 7, 11, 51, 40.905216)
longstr1 = '1 24876U 97035A   21311.49422344  .00000059  00000-0  00000-0 0  9995'
longstr2 = '2 24876  55.4827 164.3331 0054755  54.5796 306.0174  2.00562707178201'

satellite = Satrec.twoline2rv(longstr1, longstr2)

jd0, fr0 = jday(2021,11,8,22,45,0)

# Time since epoch
tsince = np.arange(0,3000,10)

# Initialization
txPos = np.zeros((len(tsince),3))
txV = np.zeros((len(tsince),3))
i = 0
jd0 = jd0 + fr0
for t in tsince:
    jd_temp = jd0 + t/1440
    jd, fr = divmod(jd_temp, 1)
    _, txPos[i,:], txV[i,:] = satellite.sgp4(jd, fr)
    i+=1
 
import matplotlib.pyplot as plt   
plt.figure(1)
plt.plot(tsince,txPos[:,0])
plt.xlabel('Time since [min]')
plt.ylabel('X coordinates [km]')
plt.title('Coordinates changes by time')   
```
============================================
[Previous question, solved]
And when I plot the coordinates change in 200 min from epoch, I can see that the coordinates change more and more rapid by time.
![Figure 2022-03-02 221944](https://user-images.githubusercontent.com/87004784/156369594-32bb9b90-fe12-4c29-8d0a-b06d7fc5d78e.png)
```
import numpy as np
from sgp4.api import jday, Satrec


longstr1 = '1 24876U 97035A   21311.49422344  .00000059  00000-0  00000-0 0  9995'
longstr2 = '2 24876  55.4827 164.3331 0054755  54.5796 306.0174  2.00562707178201'

satellite = Satrec.twoline2rv(longstr1, longstr2)

jd, fr = jday(2021,11,8,22,45,0)

# Time since epoch
tsince = np.arange(0,200,1)

# Initialization
txPos = np.zeros((len(tsince),3))
txV = np.zeros((len(tsince),3))
i = 0

for t in tsince:
    jd_temp = jd + fr + t/1440
    jd, fr = divmod(jd_temp, 1)
    _, txPos[i,:], txV[i,:] = satellite.sgp4(jd, fr)
    i+=1

import matplotlib.pyplot as plt  
  
plt.figure(1)
plt.plot(tsince,txPos[:,0])
plt.xlabel('Time since [min]')
plt.ylabel('X coordinates [km]')
plt.title('Coordinates changes by time')   
```
Looking forward to your reply.

P.S. I know it's better to use SDP model for long period orbit like GPS satellite, but for my case, the accuracy should be good enough by using SGP4 model.

**Comment by brandon-rhodes** [comment]
Could you include the plot? You should be able to edit your post, position the cursor, and if you press "Ctrl-V" with the image in your desktop clipboard, then the image should get uploaded to GitHub and included.

**Reply by jinghuiF** [reply]
Thank you very much. I have uploaded the figure.

**Reply by brandon-rhodes** [reply]
Thank you! Is there any chance you could put the code back? Then the code and diagram could be compared at the same time.

**Reply by jinghuiF** [reply]
Sorry about that. It has been edited. 
Thank you very much for your help.

**Reply by jinghuiF** [reply]
I just found the weird part of the plot is because of one mistake within the code. 
I'm sorry for the confusion.
But the disagreement of the answers shown in the beginning of the discussion still remains.

**Reply by brandon-rhodes** [reply]
I'm glad your plots are looking more periodic! I thought that maybe the code was "speeding up" when making that first plot, but I had not yet figured out why.

Let’s check: what is the epoch date of the orbital elements you are using? Are they from the archive and fairly close to the date for which you are asking for a position?

**Reply by jinghuiF** [reply]
Thank you very much for the reply.

The orbital elements I used here are read from TLE I downloaded.
Epoch date of the TLE is (2021, 11, 7, 11, 51, 40, 905216, tzinfo=UTC) and the targeting date is (2021, 11, 8, 22, 45, 00, 000000, tzinfo=UTC).
I think one day difference should have no problem providing good enough result.

**Reply by dcajacob** [reply]
Where is the "real result" coming from? A GPS receiver uses a different orbit propagator for the GPS satellites. They don't use SGP4, so that may be the issue? Ironically, the propagator used in the RXs is much simpler, but fit to better ephemeris. The coefficients are transmitted by the satellites themselves, but you can get historical almanacs.

**Reply by jinghuiF** [reply]
Thank you for your help.

Yea, I know usually if people want to get a precise position of GPS satellite they will us almanac file decoded from GPS signal. But in my case, I only need a relatively rough position to proceed next step. So to save time, I just want to use SGP4 for both receivers (LEO satellite) and transmitters (GPS).

**Reply by jinghuiF** [reply]
Thank you very much for all the help and ideas. I found out the disagreement with the answer is because of the wrong format of TLE my teammate used. The output from this library is perfectly correct. 
Thank you again for all the help
